wemake_python_styleguide/visitors/ast/statements.py

```diff
@@ -264,57 +263,72 @@ def visit_ClassDef(self, node: ast.ClassDef) -> None:
     'visit_AsyncFor',
 ))
 class VariableUsedOutsideOfBlockVisitor(BaseNodeVisitor):
-    """
-    Responsible for detection of variable defined in `with` and `for` outside of it's body
-    """
+    """Detect usage of variable defined by `with` or `for` outside of block."""
 
-    def visit_variable_defying_statement(self, node: AnyVariableDefyingStatement) -> None:
+    def visit_variable_defying_statement(
+        self,
+        node: VariableDefyingStatement,
+    ) -> None:
         """
-        Checks that variable is not used after statement body
+        Checks that variable is not used after statement body.
+
+        Raises:
+            VariableUsedOutsideOfBlockViolation
+
         """
+        self._check_variable_used_outside_of_block(node)
+        self.generic_visit(node)
+
+    def _check_variable_used_outside_of_block(
+        self,
+        node: VariableDefyingStatement,
+    ) -> None:
         node_reached = False
         defined_names = self._extract_defined_variables(node)
         if not defined_names:
             return
         for statement in node.context.body:
             if node_reached:
-                self._check_statement(defined_names, statement, node)
-            if node == statement:
+                self._check_statement_uses_forbidden_variables(
+                    defined_names, statement, node,
+                )
+            if node is statement:
                 node_reached = True
-        self.generic_visit(node)
+        return
```

Comment:
I'm pretty sure you were the one who was arguing in favor of  consistent return policies (If we return once form a function we should return in every branch eventually, avoiding implicit `return None`. So that was my reasoning. I can remove it, sure.

Reply:
@makarchuk yes, you remember correctly. But the rule is not that simple. It also has a special case for "early returns". Which is the case right now. Here's the full text: https://github.com/wemake-services/wemake-python-styleguide/issues/378